### PR TITLE
Issues/new validator

### PR DESCRIPTION
**Pull request by pmlopes**
Motivation:

A new attempt to implement a validator capable of handling the latest spec

### CODE REVIEW

**Comment by pmlopes** [comment]
PR is ready for a review

Major points:

* > 4000 TCK tests passed
* support drafts 4, 7, 2019-09, 2020-12
* codegen friendly

It's not a rewrite of the module, just the validator, the DSL is untouched. (only 1 method deprecated as it was locking the DSL to the exisitng validator)

File: src/main/asciidoc/index.adoc [hunk 1]
```diff
@@ -4,13 +4,12 @@
...
-* Non blocking `$ref` resolution and caching
-* Lookup into the schema cache using {@link io.vertx.core.json.pointer.JsonPointer}
-* Synchronous and asynchronous validation
-* Ability to extend the validation tree adding new keywords and new format predicates
+* Implementation of https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-validation-00[draft 2020-12]
+* Implementation of https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-validation-02[draft 2019-09]
+* Implementation of https://datatracker.ietf.org/doc/html/draft-handrews-json-schema-validation-01[draft 7]
+* Implementation of https://datatracker.ietf.org/doc/html/draft-fge-json-schema-validation-00[draft 4]
+* Dereferencing of `$ref` resolution and caching
+* Synchronous validation
```
Comment (vietj) [inline src/main/asciidoc/index.adoc:12]: I think no need for this.

Comment (vietj) [inline src/main/asciidoc/index.adoc:12]: I think no need for this.

File: src/main/asciidoc/index.adoc [hunk 2]
```diff
@@ -39,35 +38,45 @@ dependencies {
...
+The {@link io.vertx.json.schema.JsonSchema} interface allows both types to be handled without the constant check of the
+underlying type.
 
-The available {@link io.vertx.json.schema.SchemaParser} are:
+=== SchemaRepository
 
-* {@link io.vertx.json.schema.draft201909.Draft201909SchemaParser} for Json Schema Draft 2019-09
-* {@link io.vertx.json.schema.draft7.Draft7SchemaParser} for Json Schema Draft 7
-* {@link io.vertx.json.schema.openapi3.OpenAPI3SchemaParser} for OpenAPI 3 dialect
+The {@link io.vertx.json.schema.SchemaRepository} is the component that holds {@link io.vertx.json.schema.JsonSchema}
```
Comment (vietj) [inline src/main/asciidoc/index.adoc:53]: _is the component that holds_ simply _holds_

Comment (vietj) [inline src/main/asciidoc/index.adoc:53]: _is the component that holds_ simply _holds_

File: src/main/java/io/vertx/json/schema/Draft.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+package io.vertx.json.schema;
+
+import io.vertx.codegen.annotations.VertxGen;
+
+/**
+ * Json-Schema draft.
```
Comment (vietj) [inline src/main/java/io/vertx/json/schema/Draft.java:16]: drafts

Comment (vietj) [inline src/main/java/io/vertx/json/schema/Draft.java:16]: drafts

File: src/main/java/io/vertx/json/schema/JsonSchema.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+ * A Json-Schema holder.
+ *
+ * There are 2 kinds of Json-Schema's:
+ *
+ * <ul>
+ *   <li>JSON Object based</li>
+ *   <li>Boolean based</li>
+ * </ul>
+ *
+ * This is a common interface to handle all kids of schemas.
```
Comment (vietj) [inline src/main/java/io/vertx/json/schema/JsonSchema.java:29]: typo: all kinds

Comment (vietj) [inline src/main/java/io/vertx/json/schema/JsonSchema.java:29]: typo: all kinds

File: src/main/java/io/vertx/json/schema/JsonSchema.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  }
+
+  /**
+   * Annotates the schema. An annotation is a extra key-value added to the schema that are not relevant for
+   * validation but can be used to store pre-computed state.
+   *
+   * @param key a key
+   * @param value a value
+   */
+  void annotate(String key, String value);
```
Comment (vietj) [inline src/main/java/io/vertx/json/schema/JsonSchema.java:63]: shouldn't this be fluent ?

Comment (vietj) [inline src/main/java/io/vertx/json/schema/JsonSchema.java:63]: shouldn't this be fluent ?

**Pull request merged**